blockonomics-woocommerce.php

```diff
@@ -59,9 +59,49 @@ function blockonomics_woocommerce_init()
     add_action('woocommerce_email_customer_details', 'nolo_bnomics_woocommerce_email_customer_details', 10, 1);
     add_action('admin_enqueue_scripts', 'blockonomics_load_admin_scripts' );
     add_action('restrict_manage_posts', 'filter_orders' , 20 );
+    add_filter('woocommerce_get_checkout_payment_url','update_payment_url_on_underpayments',10,2);
     add_filter('request', 'filter_orders_by_address_or_txid' ); 
     add_filter('woocommerce_payment_gateways', 'woocommerce_add_blockonomics_gateway');
     add_filter('clean_url', 'bnomics_async_scripts', 11, 1 );
+    
+    /**
+     * Redriect to the checkout page  
+     **/
+    function  update_payment_url_on_underpayments($pay_url, $order)
+    {
+        $payment_method = $order->get_payment_method();
+        $is_blockonomics = ($payment_method === 'blockonomics');
+
+        if (!$is_blockonomics) {
+            return $pay_url;
+        }
+
+        // Check the partial payments setting
+        $blockonomics = new Blockonomics();
+        $is_partial_payments_active = $blockonomics->is_partial_payments_active();
+
+        if (!$is_partial_payments_active) {
+            return $pay_url;
+        }
+
+        global $wpdb;
+        $order_id = $order->get_id();
+        $table_name = $wpdb->prefix . 'blockonomics_payments';
+        $query = $wpdb->prepare("SELECT expected_fiat,paid_fiat,currency FROM " . $table_name . " WHERE order_id = %d ", $order_id);
+        $results = $wpdb->get_results($query, ARRAY_A);
+        $paid_fiat = $blockonomics->calculate_total_paid_fiat($results);
+
+        if (!$paid_fiat) {
+            return $pay_url;
+        }
+        
+        if ($order->needs_payment()) {
```

Comment:
Let's first perform the checks on the woocommerce order object.

1. Perform checks on the Woocommerce Order object
2. Perform checks from Blockonomics settings/database

eg.
```
$payment_method = $order->get_payment_method();

if ($payment_method !== 'blockonomics' || !$order->needs_payment()) {
    return $pay_url;
}

$blockonomics = new Blockonomics();

if (!$blockonomics->is_partial_payments_active()) {
    return $pay_url;
}

$order_id = $order->get_id();
$paid_fiat = $blockonomics->get_order_paid_fiat($order_id);

if (!$paid_fiat) {
    return $pay_url;
}

return esc_url($blockonomics->get_order_checkout_url($order_id));
```

Reply:
Oh this is so good! Thanks for helping me refactor this.

Reply:
done with changes